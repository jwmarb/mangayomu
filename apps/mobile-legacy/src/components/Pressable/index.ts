import Animated from 'react-native-reanimated';
import Pressable from './Pressable';
import type { PressableProps as NativePressableProps } from 'react-native';
import { ButtonColors } from '@mangayomu/theme';
export default Pressable;
export const AnimatedPressable = Animated.createAnimatedComponent(Pressable);
export interface PressableProps extends NativePressableProps {
  // eslint-disable-next-line @typescript-eslint/ban-types
  color?: ButtonColors | (string & {});
  ripple?: boolean;
  borderless?: boolean;
  foreground?: boolean;
  rippleRadius?: number | null;
}
