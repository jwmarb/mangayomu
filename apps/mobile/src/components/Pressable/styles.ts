import { StyleSheet } from 'react-native';

export const pressableStyles = StyleSheet.create({
  container: {
    flex: 1,
  },
});
