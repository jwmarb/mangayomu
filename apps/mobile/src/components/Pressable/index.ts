import Animated from 'react-native-reanimated';
import Pressable from './Pressable';
export default Pressable;
export const AnimatedPressable = Animated.createAnimatedComponent(Pressable);
