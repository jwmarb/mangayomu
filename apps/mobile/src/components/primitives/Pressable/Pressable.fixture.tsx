import { useSelect, useValue } from 'react-cosmos/client';
import { SafeAreaView } from 'react-native-safe-area-context';
import React from 'react';
import Pressable from '@/components/primitives/Pressable';
import useModeSelect from '@/hooks/useModeSelect';
import { createStyles } from '@/utils/theme';
import useStyles from '@/hooks/useStyles';
import Text from '@/components/primitives/Text';
import useTheme from '@/hooks/useTheme';

const styles = createStyles((theme) => ({
  container: {
    flex: 1,
    backgroundColor: theme.palette.background.default,
  },
  pressable: {
    justifyContent: 'center',
    alignItems: 'center',
  },
}));

export default function PressableFixture() {
  const mode = useModeSelect();
  const style = useStyles(styles, mode);
  const theme = useTheme(mode);

  const [title] = useValue<string>('title', {
    defaultValue: 'This whole entire screen is pressable',
  });
  const [selectedRippleRadius] = useSelect<string>('ripple radius', {
    defaultValue: 'undefined',
    options: ['16', '32', 'theme', 'undefined'] as string[],
  });
  const [selectedRippleColor] = useSelect<string>('ripple color', {
    defaultValue: 'theme',
    options: ['primary', 'secondary', 'theme', 'undefined'] as string[],
  });
  const [foreground] = useSelect('foreground', {
    defaultValue: 'false',
    options: ['true', 'false'],
  });
  const [borderless] = useSelect('foreground', {
    defaultValue: 'false',
    options: ['true', 'false'],
  });

  const rippleRadius = React.useMemo(() => {
    switch (selectedRippleRadius) {
      case 'theme':
        return theme.style.borderRadius;
      case 'undefined':
        return undefined;
      default:
        return parseInt(selectedRippleRadius);
    }
  }, [selectedRippleRadius, theme.style.borderRadius]);
  const rippleColor = React.useMemo(() => {
    switch (selectedRippleColor) {
      case 'theme':
        return theme.palette.action.ripple;
      case 'undefined':
        return undefined;
      default:
        return theme.palette[
          selectedRippleColor as unknown as 'primary' | 'secondary'
        ].main;
    }
  }, [selectedRippleColor, theme.palette]);

  // eslint-disable-next-line @typescript-eslint/no-empty-function
  function handleOnPress() {}
  return (
    <SafeAreaView style={style.container}>
      <Pressable
        style={style.pressable}
        android_ripple={{
          radius: rippleRadius,
          color: rippleColor,
          foreground: JSON.parse(foreground),
          borderless: JSON.parse(borderless),
        }}
        onPress={handleOnPress}
      >
        <Text contrast={mode}>{title}</Text>
      </Pressable>
    </SafeAreaView>
  );
}
