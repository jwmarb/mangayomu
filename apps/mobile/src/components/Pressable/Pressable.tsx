import {
  Pressable as NativePressable,
  PressableAndroidRippleConfig,
  PressableProps as NativePressableProps,
} from 'react-native';
import { pressableStyles } from '@/components/Pressable/styles';
import { createThemedProps } from '@/utils/theme';
import useThemedProps from '@/hooks/useThemedProps';
import useContrast from '@/hooks/useContrast';

const themedProps = createThemedProps((theme) => ({
  androidRipple: {
    borderless: false,
    color: theme.palette.action.ripple,
    foreground: false,
  } as PressableAndroidRippleConfig,
}));

export type PressableProps = NativePressableProps & {
  contrast?: boolean;
};

/**
 * A basic Pressable component with essential theming applied to it. It is the fundamental building block of every component that requires user interaction.
 * @returns
 */
export default function Pressable(props: PressableProps) {
  const contrast = useContrast(props.contrast);
  const { androidRipple } = useThemedProps(themedProps, contrast);
  props.android_ripple = {
    ...props.android_ripple,
    ...androidRipple,
  };
  const { android_ripple, style: styleProp, ...rest } = props;
  const style: NativePressableProps['style'] =
    typeof styleProp === 'function'
      ? (state) => [pressableStyles.container, styleProp(state)]
      : [pressableStyles.container, styleProp];

  return (
    <NativePressable {...rest} style={style} android_ripple={android_ripple} />
  );
}
