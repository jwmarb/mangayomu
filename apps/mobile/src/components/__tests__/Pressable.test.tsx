import { render, userEvent } from '@testing-library/react-native';
import React from 'react';
import Pressable from '@/components/Pressable';

beforeAll(() => {
  jest.useFakeTimers();
});

afterAll(() => {
  jest.useRealTimers();
});

test('renders properly', async () => {
  const user = userEvent.setup();
  const handleOnPress = jest.fn();
  const handleOnLongPress = jest.fn();
  const tree = render(
    <Pressable onPress={handleOnPress} onLongPress={handleOnLongPress} />,
  );
  expect(tree).toMatchSnapshot();
  await user.press(tree.root);
  await user.longPress(tree.root);

  jest.runOnlyPendingTimers();
  expect(handleOnPress).toHaveBeenCalled();
  expect(handleOnLongPress).toHaveBeenCalled();
});

test('style prop rendered properly', async () => {
  const user = userEvent.setup();
  const style = jest.fn(({ pressed }: { pressed: boolean }) =>
    pressed ? { backgroundColor: 'blue' } : { backgroundColor: 'red' },
  );
  const tree = render(<Pressable style={style} />);

  expect(tree.root).toHaveStyle({ backgroundColor: 'red' });

  await user.longPress(tree.root);
  jest.runOnlyPendingTimers();

  expect(style).toHaveBeenCalled();
  expect(style).toHaveReturnedWith({ backgroundColor: 'blue' });
});
