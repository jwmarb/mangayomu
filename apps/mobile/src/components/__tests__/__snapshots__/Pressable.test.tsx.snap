// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
/>
`;
